### DISCUSSION TITLE
How to migrate/upgrade so I can use Harman Air SDK in production? (Flex 4.14.1 != Flex 4.1A)

### DISCUSSION THREAD

**Discussion by TaS-Sensys** [airsdk/Adobe-Runtime-Support#1864] [General]
I have been attempting to put 100,000+ line Adobe Air Flex application into a Long Term Support state for the last 4 months, hired
to work on this effort part time and as little as possible after the original developer was departed.

First step was virtualizing all development hardware and build environments, which use Adobe Flash Builder 4.7 and Flex 4.1A 
So I can do builds using zombie licenses and dead code. 

In parallel I've been trying to upgrade the tool-chain - I've been using Intellij Idea and the Harman Air-SDK 33.1 + Flex 4.14.1 for debugging, but there seem to be visual and behavioral deltas I cannot get rid of to make a production quality  build. I assumed it was the Air-SDK vs Adobe Flash Builder, BUT as it turns out, the big issue is FLEX 4.1A used in Adobe Flash Builder - using Adobe Flash Builder with Flex 4.14.1 returns similar problems in casting .mxml views and some graphical elements I see in Intellij + Harman Air-SDK + Flex 4.14.1. Comparative debugging with AFB and Intellij show compatible types in all graphical elements in Flex 4.1A output that allow crazy casts that blow up under Flex 4.14.1. Its scary to come across comments in code called out as "kludge" near these magical events. Easy fixes have included setting .document types to null during creation, just evil little things that were fine 2006-2009 but failed in 2015.

IS THERE somewhere a guide or resource to upgrading from Flex 4.1A to more modern versions? At Apache Flex, the oldest
I can go is 4.9.1, which also shows the newer 4.14.1 errors. 8 version delta or are these worlds apart?

Should I be using a much higher version of Flex? (highest I see is 4.16.x?)  I would love to upgrade the build tool-chain to the Harman AIR SDK and dump the Adobe Air baggage, but there seem to be fundamental deltas in behavior leveraged far too long in Flex 4.1A for me to reverse engineer at this point.

**Comment by ajwfrost** [comment]
Hi

I'm not aware of any update guide from Apache, there's the main Flex documentation which is now at https://help.adobe.com/archive/en_US/flex/using/flex_4.6_help.pdf but I suspect it won't cover the issues you're seeing.

We have come across problems in behaviour and visual effects (e.g. with lots of layouts turning into NaN values) caused by the different ActionScript compilers. I think it was related to styles and how the Flex compiler was handling these vs the newer compiler.. we have the "Flex" variant of the AIR SDK which is basically the AIR SDK without the new compiler and libraries, so that the existing Flex version of MXMLC will still be used. But we don't maintain those versions, they were all handed off to Apache a fair few years ago.

4.16.1 is the latest Flex SDK release, from 5.5 years ago now.. so you may be better off trying to get an older one rather than a newer one! I thought 4.6 was still available from Adobe's website .. the release notes are there (https://helpx.adobe.com/flash-builder/release-note/flex-4-6-sdk-release.html) but the links within them are giving 404 responses now.

My suggestion would be to ask this of the Apache Flex support group, see the link to subscribe to the users' mailing list at the bottom of https://flex.apache.org/

Plus of course, if you can simplify your code (or create a small sample) that demonstrates the problem, it may be that someone here or there would be able to see how to get it working with Flex SDK 4.16.

thanks

**Reply by TaS-Sensys** [reply]
Hey thanks for the suggestions!

I did zip up the Adobe Flex 4.1A used by AFB, install it alone and overlaid the Harmon AIR + Flex onto it, but get a persistent "unknown configuration variable 'swf-version' holding me up that was referenced in Intellij as something-not-supported-go-get-something-newer some 7 years ago.

Unless its totally dependent on old Flex, I'd rather clean it up and use the newer tool chains. 

I am mostly getting casting/coercion #1034 errors in the console, some bubbling up to turn into error popups breaking usage, but not all, which makes me think I am progressing :)
